modules/gcp/register_fqdn/main.tf

```diff
@@ -1,19 +1,75 @@
-data "google_dns_managed_zone" "zone" {
-  project = "dns-terraform-sandbox"
-   # If the dns_zone is not set, remove the first part of the FQDN and use it
-  name = replace(coalesce(var.dns_zone, replace(var.fqdn, "/^[^\\.]+\\./", "")), ".", "-")
-}
-
 data "dns_a_record_set" "tsb" {
   host = var.address
 }
 
-resource "google_dns_record_set" "tsb_fqdn" {
+## Shared DNS Zone
+
+data "google_dns_managed_zone" "shared" {
+  count = local.shared_zone ? 1 : 0
+  project = "dns-terraform-sandbox"
+  name = local.zone_name
+}
+
+resource "google_dns_record_set" "shared_fqdn" {
+  count = local.shared_zone ? 1 : 0
   project = "dns-terraform-sandbox"
-  managed_zone = data.google_dns_managed_zone.zone.name
+  managed_zone = data.google_dns_managed_zone.shared[0].name
   name = "${var.fqdn}."
   type = "A"
   ttl  = 300
 
   rrdatas = [data.dns_a_record_set.tsb.addrs[0]]
-}
\ No newline at end of file
+}
+
+## Public DNS Zone
+
+resource "google_dns_managed_zone" "public" {
+  count     = local.public_zone ? 1 : 0
+  project   = var.project_id
+  name      = local.zone_name
+  dns_name  = "${local.dns_name}."
+}
+
+resource "google_dns_record_set" "public_fqdn" {
+  count        = local.public_zone ? 1 : 0
+  project      = var.project_id
+  managed_zone = google_dns_managed_zone.public[0].name
+  name         = "${var.fqdn}."
+  type         = "A"
+  ttl          = 300
+
+  rrdatas = [data.dns_a_record_set.tsb.addrs[0]]
+}
+
+## Private DNS Zone
+
+data "google_compute_network" "tsb" {
+  count   = local.private_zone ? 1 : 0
```

Comment:
added a "count" as it is a private zone related function @nacx jfyi

Reply:
It doesn't harm to have it without the count, as this is the network where the cluster is attached to. It always exists.